### DISCUSSION TITLE
IMAGE DESKEW STREACH

### DISCUSSION THREAD

**Discussion by GanLIT** [ImageMagick/ImageMagick#4309] [Digital Image Processing]
Hi,
Does anybody know if image correction on non-oriented trapezoid images is supported.

I need to deskew the image and also stretch and fit the image oriented in predefined width and height dimensions .

Thanks in advanced
![sample](https://user-images.githubusercontent.com/91821239/135749467-fe00d5d1-bf07-40c0-8f12-3c13f18bf6b7.jpg)

**Comment by snibgo** [comment]
What version of IM, on what platform?

Your image has a JPG extension but is actually a PNG with transparency.

I think you want to move each corner to make a rectangle. Is that correct?

If so, then you can do that with "-distort perspective". You can find the corners with "-morphology", after extracting the alpha.

**Reply by GanLIT** [reply]
Hi,
I am using IM version for windows - command line.

I don't know these commands but I will search find more details and try with your suggestion.

Thanks a lot for your effort..!

**Comment by snibgo** [comment]
Windows BAT script:
```
setlocal enabledelayedexpansion

set INFILE=135749467-fe00d5d1-bf07-40c0-8f12-3c13f18bf6b7.jpg

rem Kernel is:
rem   0 0 0
rem   0 1 -
rem   0 - 1

rem Get the corners in the order: top-left, top-right, bottom-right, bottom-left.

set N=0

for /F "usebackq tokens=1,4,5 delims=(), " %%A in (`%IMG7%magick ^
  %INFILE% ^
  -alpha extract ^
  -virtual-pixel Black ^
  -colorspace Gray ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:0,0,0,0,1,-,0,-,1" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:0,0,0,-,1,0,1,-,0" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:1,-,0,-,1,0,0,0,0" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:0,-,1,0,1,-,0,0,0" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  NULL:`) do (
    if /I %%A==Gray: (
      set X[!N!]=%%B
      set Y[!N!]=%%C
      echo !N! %%B %%C
      set /A N+=1
    )
)

magick ^
  %INFILE% ^
  -distort Perspective "%X[0]%,%Y[0]%,0,0 %X[1]%,%Y[1]%,2213,0 %X[2]%,%Y[2]%,2213,656 %X[3]%,%Y[3]%,0,656" ^
  out.jpg
```
![out](https://user-images.githubusercontent.com/33812368/135766565-d82cb186-3dbc-476d-a4a8-242c70cbafeb.jpg)

**Reply by GanLIT** [reply]
Perfect..! That was exactly what I was looking for..
I tried with another image but the result was not the same. Attached the image and the result.

I don't know if the script require changing parameters per case, and it's not generic for every image. 
As the script it's too complicated for me for fine-tuning (if it's generic), any help would be much appreciated.

Thanks a lot again for your valuable help.
 
![1](https://user-images.githubusercontent.com/91821239/135970090-ce1784d1-214c-4791-a96d-0c4a261935bb.jpg)
![out](https://user-images.githubusercontent.com/91821239/135970103-546fe497-d105-42f8-8631-7a7f53720757.jpg)

**Comment by snibgo** [comment]
I set the coordinates for the output corners to be correct for your first example, so the code is specific to that size. By finding the minimum and maximum coordinates, we can generalise:

```
setlocal enabledelayedexpansion

set INFILE=135749467-fe00d5d1-bf07-40c0-8f12-3c13f18bf6b7.jpg

set INFILE=135970090-ce1784d1-214c-4791-a96d-0c4a261935bb.png


rem Kernel is:
rem   0 0 0
rem   0 1 -
rem   0 - 1

rem Get the corners in the order: top-left, top-right, bottom-right, bottom-left.

set N=0

for /F "usebackq tokens=1,4,5 delims=(), " %%A in (`%IMG7%magick ^
  %INFILE% ^
  -alpha extract ^
+write x0.png ^
  -virtual-pixel Black ^
  -colorspace Gray ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:0,0,0,0,1,-,0,-,1" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:0,0,0,-,1,0,1,-,0" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:1,-,0,-,1,0,0,0,0" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  ^( +clone ^
    -morphology HMT "3x3+1+1:0,-,1,0,1,-,0,0,0" ^
    -define "identify:limit=1" ^
    -define "identify:locate=maximum" ^
    +write info: ^
    +delete ^) ^
  NULL:`) do (
    if /I %%A==Gray: (
      set X[!N!]=%%B
      set Y[!N!]=%%C
      echo !N! %%B %%C
      set /A N+=1
    )
)

set minX=%X[0]%
if %minX% GTR %X[1]% set minX=%X[1]%
if %minX% GTR %X[2]% set minX=%X[2]%
if %minX% GTR %X[3]% set minX=%X[3]%

set minY=%Y[0]%
if %minY% GTR %Y[1]% set minY=%Y[1]%
if %minY% GTR %Y[2]% set minY=%Y[2]%
if %minY% GTR %Y[3]% set minY=%Y[3]%

set maxX=%X[0]%
if %maxX% LSS %X[1]% set maxX=%X[1]%
if %maxX% LSS %X[2]% set maxX=%X[2]%
if %maxX% LSS %X[3]% set maxX=%X[3]%

set maxY=%Y[0]%
if %maxY% LSS %Y[1]% set maxY=%Y[1]%
if %maxY% LSS %Y[2]% set maxY=%Y[2]%
if %maxY% LSS %Y[3]% set maxY=%Y[3]%

set /A maxX-=1
set /A maxY-=1

%IMG7%magick ^
  %INFILE% ^
  -distort Perspective "%X[0]%,%Y[0]%,%minX%,%minY% %X[1]%,%Y[1]%,%maxX%,%minY% %X[2]%,%Y[2]%,%maxX%,%maxY% %X[3]%,%Y[3]%,%minX%,%maxY%" ^
  +repage ^
  out.png
```

**Reply by GanLIT** [reply]
That's much better.

Thanks again a lot for your effort..!

**Comment by GanLIT** [comment]
Hi, I wish all the best for a happy and healthy New Year,

Is there a way to convert the script in .net ? For some reason, it can't be executed it through .net environment using process.start & waitforexit as process stops before completion...! 
Also for some images like the one below, the script never ends... Is there a way to stop execution if process fails?

![9C](https://user-images.githubusercontent.com/91821239/147939356-80817fb0-fac9-4ef3-9559-e672639d3da2.JPG)

Any help would be mush appreciated 
Thanks again **A LOT** for your help.
